test/common/network/connection_impl_test.cc

```diff
@@ -264,6 +330,133 @@ TEST_P(ConnectionImplTest, BasicWrite) {
+  Buffer::OwnedImpl first_buffer_to_write(data_to_write);
+  std::string data_written;
+  EXPECT_CALL(*client_write_buffer_, move(_))
+      .WillRepeatedly(DoAll(AddBufferToStringWithoutDraining(&data_written),
+                            Invoke(client_write_buffer_, &MockBuffer::baseMove)));
+  EXPECT_CALL(*client_write_buffer_, write(_))
+      .WillOnce(Invoke(client_write_buffer_, &MockBuffer::trackWrites));
+  // The write() call on the connection will buffer enough data to bring the connection above the
+  // high watermark but the subsequent drain immediately brings it back below.
+  // A nice future performance optimization would be to latch if the socket is writable in the
+  // connection_impl, and try an immediate drain inside of write() to avoid thrashing here.
+  EXPECT_CALL(client_callbacks_, onAboveWriteBufferHighWatermark());
+  EXPECT_CALL(client_callbacks_, onBelowWriteBufferLowWatermark());
+  client_connection_->write(first_buffer_to_write);
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  EXPECT_EQ(data_to_write, data_written);
+
+  // Now do the write again, but this time configure buffer_ to reject the write
+  // with errno set to EAGAIN via failWrite().  This should result in going above the high
+  // watermark and not returning.
+  Buffer::OwnedImpl second_buffer_to_write(data_to_write);
+  EXPECT_CALL(*client_write_buffer_, move(_))
+      .WillRepeatedly(DoAll(AddBufferToStringWithoutDraining(&data_written),
+                            Invoke(client_write_buffer_, &MockBuffer::baseMove)));
+  EXPECT_CALL(*client_write_buffer_, write(_))
+      .WillOnce(Invoke(client_write_buffer_, &MockBuffer::failWrite));
+  // The write() call on the connection will buffer enough data to bring the connection above the
+  // high watermark and as the data will not flush it should not return below the watermark.
+  EXPECT_CALL(client_callbacks_, onAboveWriteBufferHighWatermark());
+  EXPECT_CALL(client_callbacks_, onBelowWriteBufferLowWatermark()).Times(0);
+  client_connection_->write(second_buffer_to_write);
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+
+  // Clean up the connection.  The close() will attempt to flush.  The call to
+  // write() will succeed, bringing the connection back under the low watermark.
+  EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::LocalClose));
+  EXPECT_CALL(*client_write_buffer_, write(_))
+      .WillOnce(Invoke(client_write_buffer_, &MockBuffer::trackWrites));
+  EXPECT_CALL(client_callbacks_, onBelowWriteBufferLowWatermark()).Times(1);
+  client_connection_->close(ConnectionCloseType::NoFlush);
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+}
+
+// Read and write random bytes and ensure we don't encounter issues.
+TEST_P(ConnectionImplTest, WatermarkFuzzing) {
+  useMockBuffer();
+  setUpBasicConnection();
+
+  connect();
+  client_connection_->setBufferLimits(10);
+
+  Runtime::RandomGeneratorImpl rand;
```

Comment:
Can we use a seed that can be printed and supplied on the command-line to repeat failed tests? I worry about failures we catch in CI flakes that we can't reproduce otherwise. Maybe the gtest seed (https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#shuffling-the-tests) ?